src/test/parallel.py

```diff
@@ -0,0 +1,864 @@
+    t.start()
+    return t
+
+  try:
+    if timeout:
+      timeout.start()
+    running_groups = set() if serialize_test_cases else None
+    worker_fn = WorkerFn(tasks, running_groups)
+    workers = [start_daemon(worker_fn) for _ in range(pool_size)]
+    for worker in workers:
+      worker.join()
+  finally:
+    if timeout:
+      timeout.cancel()
+
+
+def default_options_parser(default_binary, file_test_list):
+  parser = optparse.OptionParser(
+      usage = 'usage: %prog [options] binary [binary ...] -- [additional args]')
+
+  parser.add_option('-d', '--output_dir', type='string', default=None,
+                    help='Output directory for test logs. Logs will be '
+                         'available under gtest-parallel-logs/, so '
+                         '--output_dir=/tmp will results in all logs being '
+                         'available under /tmp/gtest-parallel-logs/.')
+  parser.add_option('-r', '--repeat', type='int', default=1,
+                    help='Number of times to execute all the tests.')
+  parser.add_option('--retry_failed', type='int', default=0,
+                    help='Number of times to repeat failed tests.')
+  parser.add_option('--failed', action='store_true', default=False,
+                    help='run only failed and new tests')
+  parser.add_option('-w', '--workers', type='int',
+                    default=multiprocessing.cpu_count(),
+                    help='number of workers to spawn')
+  parser.add_option('--gtest_color', type='string', default='yes',
+                    help='color output')
+  parser.add_option('--gtest_filter', type='string', default='',
+                    help='test filter')
+  parser.add_option('--gtest_also_run_disabled_tests', action='store_true',
+                    default=False, help='run disabled tests too')
+  parser.add_option('--print_test_times', action='store_true', default=False,
+                    help='list the run time of each test at the end of execution')
+  parser.add_option('--shard_count', type='int', default=1,
+                    help='total number of shards (for sharding test execution '
+                         'between multiple machines)')
+  parser.add_option('--shard_index', type='int', default=0,
+                    help='zero-indexed number identifying this shard (for '
+                         'sharding test execution between multiple machines)')
+  parser.add_option('--dump_json_test_results', type='string', default=None,
+                    help='Saves the results of the tests as a JSON machine-'
+                         'readable file. The format of the file is specified at '
+                         'https://www.chromium.org/developers/the-json-test-results-format')
+  parser.add_option('--timeout', type='int', default=None,
+                    help='Interrupt all remaining processes after the given '
+                         'time (in seconds).')
+  parser.add_option('--serialize_test_cases', action='store_true',
```

Comment:
I think this option is currently incompatible with sharding, since it shards tests in a round-robin type fashion, thus splitting up test_groups between shards. This is probably desired, but just need to document it, or try to prevent the two options being used in combination.